File: moling-server/app/service/health_service.py
"""墨灵 (Moling) — Health Service.

提供项目健康检查逻辑：R1 角色一致性、R2 时间线连续性、R3 伏笔债务。
"""

from __future__ import annotations

import json
import logging
from datetime import datetime, timezone
from typing import Optional, Dict, List, Any

from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import get_settings
from app.dao import project_dao, vault_dao
from app.errors import NotFoundError, ErrorCode, AppError
from app.models.health_alert import HealthAlert
from app.models.chapter import Chapter
from app.models.project import Project
from app.models.vault_character import VaultCharacter
from app.models.vault_plot_promise import VaultPlotPromise
from app.models.vault_timeline import VaultTimeline
from app.models.vault_world import VaultWorld
from app.llm.client import llm_client

logger = logging.getLogger(__name__)
settings = get_settings()


class HealthService:
    """Service for project health checks (R1/R2/R3)."""

    async def run_health_check(
        self,
        db: AsyncSession,
        user_id: str,
        project_id: int,
    ) -> Dict[str, Any]:
        """Run all health checks (R1/R2/R3) for a project.
        
        Args:
            db: Database session
            user_id: User ID
            project_id: Project ID
            
        Returns:
            Health check results with alerts
        """
        # Verify project exists and belongs to user
        project = await project_dao.get(db, project_id)
        if project is None:
            raise NotFoundError(
                error_code=ErrorCode.PROJECT_NOT_FOUND,
                detail="Project not found",
            )
        if project.user_id != user_id:
            raise AppError(
                error_code=ErrorCode.PROJECT_ACCESS_DENIED,
                detail="Not authorized to access this project",
            )

        result = {
            "project_id": project_id,
            "checked_at": datetime.now(timezone.utc).isoformat(),
            "checks": {},
            "alerts": [],
        }

        # ===== R1: Character Consistency Check =====
        logger.info(f"Health check R1: Character consistency for project {project_id}")
        r1_result = await self._check_r1_character_consistency(db, project)
        result["checks"]["R1"] = r1_result

        # ===== R2: Timeline Continuity Check =====
        logger.info(f"Health check R2: Timeline continuity for project {project_id}")
        r2_result = await self._check_r2_timeline_continuity(db, project)
        result["checks"]["R2"] = r2_result

        # ===== R3: Plot Promise Debt Check =====
        logger.info(f"Health check R3: Plot promise debt for project {project_id}")
        r3_result = await self._check_r3_plot_promise_debt(db, project)
        result["checks"]["R3"] = r3_result

        # Collect alerts
        for check_key, check_result in result["checks"].items():
            if not check_result["passed"]:
                alerts = check_result.get("alerts", [])
                for alert in alerts:
                    # Create or update HealthAlert record
                    health_alert = HealthAlert(
                        project_id=project_id,
                        rule=check_key,
                        title=alert.get("title", "Health Issue"),
                        detail=alert.get("detail", ""),
                        severity=alert.get("severity", "warning"),
                        is_active=True,
                        checked_at=datetime.now(timezone.utc),
                    )
                    db.add(health_alert)
                    result["alerts"].append(alert)

        await db.commit()

        logger.info(f"Health check completed for project {project_id}: {len(result['alerts'])} alerts")
        return result

    async def _check_r1_character_consistency(
        self,
        db: AsyncSession,
        project: Project,
    ) -> Dict[str, Any]:
        """R1: Check character behavior consistency.
        
        Checks:
        - Are character traits consistent across chapters?
        - Are character relationships logically maintained?
        - Are there any character "personality conflicts"?
        """
        result = {
            "passed": True,
            "score": 1.0,
            "details": "",
            "alerts": [],
        }

        try:
            # Get all characters
            characters = await vault_dao.get_characters(db, project.id)
            
            if not characters:
                result["passed"] = True
                result["score"] = 1.0
                result["details"] = "No characters to check"
                return result

            # Check for character inconsistencies using LLM
            character_data = []
            for char in characters:
                character_data.append({
                    "name": char.name,
                    "role": char.role,
                    "traits": char.traits,
                    "emotion": char.emotion,
                    "chapter_count": char.chapter_count,
                })

            prompt = f"""请分析以下小说角色的一致性健康状况。

项目：{project.title}

角色列表：
{json.dumps(character_data, ensure_ascii=False, indent=2)}

请检查：
1. 角色特征是否清晰且一致？
2. 角色出场章节数是否合理？
3. 是否有角色"personality conflicts"？

返回 JSON 格式：
{{"consistent": true/false, "issues": ["问题1", "问题2"], "score": 0.0-1.0, "recommendations": ["建议1", "建议2"]}}
"""

            response_text = await self._call_llm(prompt)
            
            try:
                llm_result = json.loads(response_text)
                
                result["passed"] = llm_result.get("consistent", True)
                result["score"] = llm_result.get("score", 0.9)
                result["details"] = f"角色一致性得分: {result['score']:.2f}"
                
                if not result["passed"]:
                    issues = llm_result.get("issues", [])
                    recommendations = llm_result.get("recommendations", [])
                    
                    for issue in issues:
                        result["alerts"].append({
                            "rule": "R1",
                            "title": "角色一致性问题",
                            "detail": issue,
                            "severity": "warning",
                            "recommendation": ", ".join(recommendations) if recommendations else "",
                        })
                
            except json.JSONDecodeError:
                # If LLM response is not valid JSON, use default
                result["passed"] = True
                result["score"] = 0.85
                result["details"] = "角色一致性检查完成（LLM 响应解析失败）"

        except Exception as e:
            logger.error(f"R1 check failed: {e}", exc_info=True)
            result["passed"] = True
            result["score"] = 0.8
            result["details"] = f"检查失败: {str(e)}"

        return result

    async def _check_r2_timeline_continuity(
        self,
        db: AsyncSession,
        project: Project,
    ) -> Dict[str, Any]:
        """R2: Check timeline continuity.
        
        Checks:
        - Are timeline events in correct chronological order?
        - Are there any "time gaps" or "time conflicts"?
        - Do character appearances match timeline?
        """
        result = {
            "passed": True,
            "score": 1.0,
            "details": "",
            "alerts": [],
        }

        try:
            # Get all timeline events
            events = await vault_dao.get_timeline(db, project.id)
            
            if not events or len(events) < 2:
                result["passed"] = True
                result["score"] = 1.0
                result["details"] = "Timeline events insufficient for continuity check"
                return result

            # Check timeline order
            timeline_data = []
            for event in events:
                timeline_data.append({
                    "chapter_number": event.chapter_number,
                    "event": event.event,
                    "description": event.description,
                    "is_key_event": event.is_key_event,
                    "characters_involved": event.characters_involved,
                })

            # Simple check: are events in chapter order?
            chapter_numbers = [e.chapter_number for e in events]
            if chapter_numbers != sorted(chapter_numbers):
                result["passed"] = False
                result["score"] = 0.6
                result["details"] = "时间线事件章节顺序不一致"
                result["alerts"].append({
                    "rule": "R2",
                    "title": "时间线顺序问题",
                    "detail": "时间线事件的章节顺序不一致，可能存在时间线冲突",
                    "severity": "warning",
                })
            else:
                result["passed"] = True
                result["score"] = 0.95
                result["details"] = "时间线连续性检查通过"

        except Exception as e:
            logger.error(f"R2 check failed: {e}", exc_info=True)
            result["passed"] = True
            result["score"] = 0.8
            result["details"] = f"检查失败: {str(e)}"

        return result

    async def _check_r3_plot_promise_debt(
        self,
        db: AsyncSession,
        project: Project,
    ) -> Dict[str, Any]:
        """R3: Check plot promise debt.
        
        Checks:
        - Are there too many unresolved plot promises (debt)?
        - Are there "abandoned" promises?
        - Are promises being resolved at a reasonable rate?
        """
        result = {
            "passed": True,
            "score": 1.0,
            "details": "",
            "alerts": [],
        }

        try:
            # Get all plot promises
            promises = await vault_dao.get_plot_promises(db, project.id)
            
            if not promises:
                result["passed"] = True
                result["score"] = 1.0
                result["details"] = "No plot promises to check"
                return result

            # Categorize promises by status
            dormant_count = len([p for p in promises if p.status == "dormant"])
            active_count = len([p for p in promises if p.status == "active"])
            resolved_count = len([p for p in promises if p.status == "resolved"])
            abandoned_count = len([p for p in promises if p.status == "abandoned"])

            total = len(promises)
            resolved_rate = resolved_count / total if total > 0 else 0

            # Check debt: too many unresolved promises?
            unresolved_count = dormant_count + active_count
            debt_threshold = 10  # Configurable threshold

            if unresolved_count > debt_threshold:
                result["passed"] = False
                result["score"] = 0.5
                result["details"] = f"伏笔债务过高: {unresolved_count} 个未回收伏笔"
                result["alerts"].append({
                    "rule": "R3",
                    "title": "伏笔债务警告",
                    "detail": f"当前有 {unresolved_count} 个未回收的伏笔，建议尽快回收部分伏笔",
                    "severity": "critical",
                    "unresolved_count": unresolved_count,
                    "resolved_rate": resolved_rate,
                })
            elif unresolved_count > debt_threshold * 0.7:
                result["passed"] = True
                result["score"] = 0.7
                result["details"] = f"伏笔债务偏高: {unresolved_count} 个未回收伏笔"
                result["alerts"].append({
                    "rule": "R3",
                    "title": "伏笔债务提醒",
                    "detail": f"当前有 {unresolved_count} 个未回收的伏笔，建议关注伏笔回收",
                    "severity": "warning",
                    "unresolved_count": unresolved_count,
                    "resolved_rate": resolved_rate,
                })
            else:
                result["passed"] = True
                result["score"] = 0.9
                result["details"] = f"伏笔债务健康: {unresolved_count} 个未回收伏笔，回收率 {resolved_rate:.1%}"

        except Exception as e:
            logger.error(f"R3 check failed: {e}", exc_info=True)
            result["passed"] = True
            result["score"] = 0.8
            result["details"] = f"检查失败: {str(e)}"

        return result

    async def get_alerts(
        self,
        db: AsyncSession,
        project_id: int,
        active_only: bool = True,
    ) -> list[HealthAlert]:
        """Get health alerts for a project."""
        stmt = select(HealthAlert).where(
            HealthAlert.project_id == project_id
        )
        if active_only:
            stmt = stmt.where(HealthAlert.is_active == True)
        stmt = stmt.order_by(HealthAlert.created_at.desc())
        result = await db.execute(stmt)
        return list(result.scalars().all())

    async def _call_llm(self, prompt: str) -> str:
        """Call LLM with prompt and return response text."""
        messages = [
            {"role": "system", "content": "你是一个专业的小说健康分析助手。"},
            {"role": "user", "content": prompt},
        ]
        
        response = await llm_client.chat(
            messages=messages,
            model=settings.LLM_MODEL,
            temperature=0.3,
            max_tokens=2048,
        )
        
        return response["choices"][0]["message"]["content"]


health_service = HealthService()

File: moling-server-rs/crates/moling-services/src/health_service.rs
//! Health service — system health check and monitoring (R1/R2/R3).
//!
//! Mirrors Python `app/service/health_service.py`.
//!
//! R1: Character consistency check
//! R2: Timeline continuity check
//! R3: Plot promise debt check

use moling_core::error::{AppError, AppResult};
use moling_db::dao::health_alert_dao::HealthAlertDao;
use moling_db::dao::project_dao::ProjectDao;
use moling_db::dao::vault_dao::VaultDao;
use moling_db::entities::health_alert::Model as HealthAlertModel;
use sea_orm::{DatabaseConnection, Set};

/// System health status.
#[derive(Clone)]
pub struct HealthService {
    health_alert_dao: HealthAlertDao,
    vault_dao: VaultDao,
    project_dao: ProjectDao,
}

impl HealthService {
    pub fn new() -> Self {
        Self {
            health_alert_dao: HealthAlertDao,
            vault_dao: VaultDao,
            project_dao: ProjectDao,
        }
    }

    // ---- Infrastructure checks ----

    /// Check database connectivity.
    pub async fn check_db(&self, db: &DatabaseConnection) -> AppResult<bool> {
        db.ping().await.map(|_| true).or(Ok(false))
    }

    /// Check Redis connectivity (delegates to pool health).
    pub fn check_redis(&self, redis_pool: &moling_core::redis::RedisPool) -> bool {
        redis_pool.is_available()
    }

    /// Full health check: database + Redis.
    pub async fn full_check(
        &self,
        db: &DatabaseConnection,
        redis_pool: &moling_core::redis::RedisPool,
    ) -> AppResult<HealthStatus> {
        let db_ok = self.check_db(db).await?;
        Ok(HealthStatus {
            database: db_ok,
            redis: self.check_redis(redis_pool),
            status: if db_ok { "healthy" } else { "degraded" }.to_owned(),
        })
    }

    // ---- R1/R2/R3 Project Health Checks ----

    /// Run all health checks (R1/R2/R3) for a project.
    ///
    /// Mirrors Python `HealthService.run_health_check`.
    pub async fn run_health_check(
        &self,
        db: &DatabaseConnection,
        user_id: &str,
        project_id: i32,
    ) -> AppResult<serde_json::Value> {
        // Verify project ownership
        let project = self.project_dao.find_by_id(db, project_id)
            .await?
            .ok_or_else(AppError::project_not_found)?;
        if project.user_id != user_id {
            return Err(AppError::project_access_denied());
        }

        let mut result = serde_json::json!({
            "project_id": project_id,
            "checks": {},
            "alerts": [],
        });

        // R1: Character consistency
        let r1 = self.check_r1(db, project_id).await?;
        result["checks"]["R1"] = serde_json::json!({
            "passed": r1.passed,
            "score": r1.score,
            "details": r1.details,
        });

        // R2: Timeline continuity
        let r2 = self.check_r2(db, project_id).await?;
        result["checks"]["R2"] = serde_json::json!({
            "passed": r2.passed,
            "score": r2.score,
            "details": r2.details,
        });

        // R3: Plot promise debt
        let r3 = self.check_r3(db, project_id).await?;
        result["checks"]["R3"] = serde_json::json!({
            "passed": r3.passed,
            "score": r3.score,
            "details": r3.details,
        });

        // Collect and create alerts for failed checks
        let mut alerts = Vec::new();
        let checks = [(r1, "R1"), (r2, "R2"), (r3, "R3")];
        for (check, rule) in &checks {
            if !check.passed {
                for alert_detail in &check.alert_details {
                    let alert = self.create_alert(
                        db, project_id, rule,
                        &alert_detail.title,
                        &alert_detail.detail,
                        &alert_detail.severity,
                    ).await?;
                    alerts.push(serde_json::json!({
                        "id": alert.id,
                        "rule": alert.rule,
                        "title": alert.title,
                        "severity": alert.severity,
                    }));
                }
            }
        }
        result["alerts"] = serde_json::json!(alerts);

        Ok(result)
    }

    /// R1: Check character behavior consistency.
    async fn check_r1(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
    ) -> AppResult<CheckResult> {
        let characters = self.vault_dao.find_characters(db, project_id).await?;

        if characters.is_empty() {
            return Ok(CheckResult {
                passed: true,
                score: 1.0,
                details: "No characters to check".into(),
                alert_details: vec![],
            });
        }

        // Check for consistency issues
        let mut alert_details = Vec::new();
        let mut score: f64 = 1.0;
        // Check: characters with very low chapter count
        let low_appearance: Vec<_> = characters
            .iter()
            .filter(|c| c.chapter_count < 2 && characters.len() > 5)
            .collect();
        if !low_appearance.is_empty() {
            score = (score - 0.15_f64).max(0.0);
            alert_details.push(AlertDetail {
                title: "低出场率角色".into(),
                detail: format!(
                    "{} 个角色出场次数不足，可能被遗忘",
                    low_appearance.len()
                ),
                severity: "warning".into(),
            });
        }

        // Check: duplicate character names
        let names: Vec<&str> = characters.iter().map(|c| c.name.as_str()).collect();
        let mut seen = std::collections::HashSet::new();
        let mut duplicates = Vec::new();
        for name in &names {
            if !seen.insert(*name) {
                duplicates.push(*name);
            }
        }
        if !duplicates.is_empty() {
            score = (score - 0.2_f64).max(0.0);
            alert_details.push(AlertDetail {
                title: "重复角色名".into(),
                detail: format!("检测到重复角色名: {}", duplicates.join(", ")),
                severity: "critical".into(),
            });
        }

        Ok(CheckResult {
            passed: score >= 0.7,
            score: score.max(0.0),
            details: format!("角色一致性得分: {:.2}", score),
            alert_details,
        })
    }

    /// R2: Check timeline continuity.
    async fn check_r2(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
    ) -> AppResult<CheckResult> {
        let events = self.vault_dao.find_timeline_events(db, project_id).await?;

        if events.len() < 2 {
            return Ok(CheckResult {
                passed: true,
                score: 1.0,
                details: "Timeline events insufficient for continuity check".into(),
                alert_details: vec![],
            });
        }

        let mut alert_details = Vec::new();
        let mut score: f64 = 1.0;
        // Check: events in chronological order by chapter_number
        let chapter_numbers: Vec<i32> = events.iter().map(|e| e.chapter_number).collect();
        let mut sorted = chapter_numbers.clone();
        sorted.sort();

        if chapter_numbers != sorted {
            score = (score - 0.4_f64).max(0.0);
            alert_details.push(AlertDetail {
                title: "时间线顺序问题".into(),
                detail: "时间线事件的章节顺序不一致，可能存在时间线冲突".into(),
                severity: "warning".into(),
            });
        }

        Ok(CheckResult {
            passed: score >= 0.7,
            score: score.max(0.0),
            details: if score >= 0.95 {
                "时间线连续性检查通过".into()
            } else {
                format!("时间线连续性得分: {:.2}", score)
            },
            alert_details,
        })
    }

    /// R3: Check plot promise debt.
    async fn check_r3(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
    ) -> AppResult<CheckResult> {
        let promises = self.vault_dao.find_plot_promises(db, project_id).await?;

        if promises.is_empty() {
            return Ok(CheckResult {
                passed: true,
                score: 1.0,
                details: "No plot promises to check".into(),
                alert_details: vec![],
            });
        }

        let total = promises.len();
        let dormant_count = promises.iter().filter(|p| p.status == "dormant").count();
        let active_count = promises.iter().filter(|p| p.status == "active").count();
        let resolved_count = promises.iter().filter(|p| p.status == "resolved").count();
        let unresolved = dormant_count + active_count;
        let resolved_rate = resolved_count as f64 / total as f64;

        let debt_threshold: usize = 10;

        let mut alert_details = Vec::new();
        let mut score: f64 = 1.0;        let mut passed = true;

        if unresolved > debt_threshold {
            score = 0.5;
            passed = false;
            alert_details.push(AlertDetail {
                title: "伏笔债务警告".into(),
                detail: format!(
                    "当前有 {unresolved} 个未回收的伏笔，建议尽快回收部分伏笔"
                ),
                severity: "critical".into(),
            });
        } else if unresolved as f64 > debt_threshold as f64 * 0.7 {
            score = 0.7;
            alert_details.push(AlertDetail {
                title: "伏笔债务提醒".into(),
                detail: format!(
                    "当前有 {unresolved} 个未回收的伏笔，建议关注伏笔回收"
                ),
                severity: "warning".into(),
            });
        }

        Ok(CheckResult {
            passed,
            score: score.max(0.0),
            details: if passed {
                format!("伏笔债务健康: {unresolved} 个未回收伏笔，回收率 {:.1}%", resolved_rate * 100.0)
            } else {
                format!("伏笔债务过高: {unresolved} 个未回收伏笔")
            },
            alert_details,
        })
    }

    // ---- Alert management ----

    /// Create a health alert.
    pub async fn create_alert(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
        rule: &str,
        title: &str,
        detail: &str,
        severity: &str,
    ) -> AppResult<HealthAlertModel> {
        let model = moling_db::entities::health_alert::ActiveModel {
            project_id: Set(project_id),
            rule: Set(rule.to_owned()),
            title: Set(title.to_owned()),
            detail: Set(detail.to_owned()),
            severity: Set(severity.to_owned()),
            is_active: Set(true),
            ..Default::default()
        };
        self.health_alert_dao.create(db, model).await
    }

    /// List health alerts for a project.
    ///
    /// Mirrors Python `HealthService.get_alerts`.
    pub async fn list_alerts(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
        active_only: bool,
    ) -> AppResult<Vec<HealthAlertModel>> {
        if active_only {
            self.health_alert_dao.list_active_by_project(db, project_id).await
        } else {
            self.health_alert_dao.list_by_project(db, project_id).await
        }
    }

    /// List alerts filtered by severity level.
    ///
    /// Severity levels: "info", "warning", "critical".
    pub async fn list_alerts_by_severity(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
        severity: &str,
    ) -> AppResult<Vec<HealthAlertModel>> {
        self.health_alert_dao.list_by_severity(db, project_id, severity).await
    }

    /// Auto-resolve all active alerts for a given rule in a project.
    ///
    /// Called when a health check passes after previously failing.
    pub async fn resolve_by_rule(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
        rule: &str,
    ) -> AppResult<u64> {
        self.health_alert_dao.resolve_by_rule(db, project_id, rule).await
    }

    /// Update the checked_at timestamp for an alert.
    pub async fn touch_alert(
        &self,
        db: &DatabaseConnection,
        alert_id: i32,
    ) -> AppResult<()> {
        self.health_alert_dao.update_checked_at(db, alert_id).await
    }

    /// Get health monitor summary for a project.
    pub async fn monitor(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
        user_id: &str,
    ) -> AppResult<serde_json::Value> {
        // Run all checks
        let check_result = self.run_health_check(db, user_id, project_id).await?;

        // Get all active alerts
        let active_alerts = self.list_alerts(db, project_id, true).await?;
        let critical_count = active_alerts.iter().filter(|a| a.severity == "critical").count();
        let warning_count = active_alerts.iter().filter(|a| a.severity == "warning").count();

        Ok(serde_json::json!({
            "project_id": project_id,
            "status": if critical_count > 0 { "critical" } else if warning_count > 0 { "warning" } else { "healthy" },
            "alert_counts": {
                "total": active_alerts.len(),
                "critical": critical_count,
                "warning": warning_count,
            },
            "last_check": check_result,
        }))
    }
}

impl Default for HealthService {
    fn default() -> Self {
        Self::new()
    }
}

/// System health response.
pub struct HealthStatus {
    pub status: String,
    pub database: bool,
    pub redis: bool,
}

/// Internal check result for R1/R2/R3.
struct CheckResult {
    passed: bool,
    score: f64,
    details: String,
    alert_details: Vec<AlertDetail>,
}

/// Detail for an alert within a check result.
struct AlertDetail {
    title: String,
    detail: String,
    severity: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_health_service_constructs() {
        let _ = HealthService::new();
    }

    #[test]
    fn test_health_status() {
        let status = HealthStatus {
            status: "healthy".into(),
            database: true,
            redis: true,
        };
        assert_eq!(status.status, "healthy");
        assert!(status.database);
    }
}
